Federation Examples (#431)
samsarahq/thunder: federationexample/README.md

# Graphql Federation

> In our current world, we have a single graphql schema that contains all theh obejcts and field funcs that need to run. However, as the number of objects and field funcs scale, we want to allow product teams to each have their own gqlserver that conatain the portion of the gqlschema relevant to the product team.

Reviewer: ```suggestion
In our current world, we have a single graphql schema that contains all the obejcts and field funcs that need to be executed at the same server (gqlexecutorservice). However, as the number of objects and field funcs scale, we want to allow product teams to each have their own gqlserver that contains the portion of the gqlschema relevant to the product team.
```

> Federation helps by creating a service called the gqlgateway that is aware of every team gqlserver and it's schemas. Queries are now sent to the gateway whichh parses and sends subqueries to each of the gqlservers. The subqueries only contains field funcs that the team's gqlserver know's how to resolve. After sending all the subqueries, it stitches the results into a single result sent back to the client. This ensures the client can query any field while the schema is federated under the hood. 

Reviewer: ```suggestion
Federation helps by creating a service called the gqlgateway that is aware of each gqlserver, that belongs to individual team, and its associated schemas. Queries are parsed at the main gateway and broken down to sub-queries, which are forwarded to the appropriate gqlservers for execution. The sub-queries only contains field funcs that each gqlserver is able to resolve. After receiving all results from the sub-queries, the main gateway stitches all the results into a single result to respond to the original query from the client. This ensures the client can query any field while the schema is federated under the hood. 
```

> We want to support this to help in terms of giving teams more ownership over their portion of the graphql schema along and with giving team's isolation such thhat there work with their portion of the graphhql schema doesn't effect other teams.

Reviewer: ```suggestion
Federation will provide teams more ownership over graphql schema and isolate executor to provide more operational stability.
```

## Architecture Diagram

![Screenshot](federation.png)

## API 

> If an object entirely belongs to a gqlserver, it can exist on only that gqlserver with all it's field funcs. However, there are some objects that have fields that belong to different gqlservers. We call these federated objects. 

Reviewer: ```suggestion
If an object entirely belongs to a gqlserver, it can exist on only that gqlserver with all its field funcs. However, there are some objects that have fields that belong to multiple gqlservers. We call these federated objects. 
```

> A nonfederated object looks something like this example below 

Reviewer: ```suggestion
A non-federated object looks something like this example below:
```

```
type Device struct {
    Id            int64
    OrgId         int64
    Name          string
}
device := schema.Object("Device", Device{})
```

> A federated object has an extra field called FetchObjectFromKeys that takes a function, telling the gqlserver how to fecth the full object from a the federated keys, the minimum number of fields needed to fetch the object. A federated object can exist on multiple gqlservers and hhave different field funcs on each of the team's gqlservers. 

Reviewer: ```suggestion
A federated object has an extra field called FetchObjectFromKeys that takes a function, telling the gqlserver how to fetch the full object using federated keys; the minimum number of fields required to fetch the object. A federated object can exist on multiple gqlservers and have different field funcs on each gqlserver. 
```

```
type Device struct {
    Id            int64
    OrgId         int64
    Name          string
}
// Device keys has to be a subset of the fields on Device and is the minimum information needded to fetch a full Device
type DeviceKeys struct {
    Id            int64
    OrgId         int64
}
device := schema.Object("Device", Device{}, schemabuilder.FetchObjectFromKeys(func(args struct{ Keys []*DeviceKeys }) []*Device {
    // Fetch device from database using device keys    
    return fetchedDevices
}))

> ```

Reviewer: This example is a bit confusing to me... can we work on this together to provide a bit more descriptive and detailed example of `federated object`?

Author: Sounds good! Let's find some time to talk about it tomorrow

## Example

> In teh federation example, we have 2 gqlservers (device and safety). Device and Location are federated objects, whcih allows them to have field funcs on both gqlservers. 

Reviewer: ```suggestion
In this example, we have 2 gqlservers (device and safety). There are also two federated objects, Device and Location, which have field funcs on both gqlservers. 
```

```
{
  device {                  #devicegqlserver
    id                      #devicegqlserver
    location {              #devicegqlserver
      latitude              #devicegqlserver
      longitude             #devicegqlserver
      street                #devicegqlserver
    }
    camera{                 #safetygqlserver
      id                    #safetygqlserver
      isOn                  #safetygqlserver
      isMultiCam            #safetygqlserver
      cameraLocation {      #safetygqlserver
      	latitude            #safetygqlserver
      	longitude           #safetygqlserver
        street              #devicegqlserver
   	  }
    }
  }
}

> ```

Reviewer: I feel that `camera` came out of no where at first glance.  Can we also have description on RE: `camera`?

Author: I added more explanation at the beginning of what each gqlserver is supposed to own and why the camera is on safetygqlserver

> When the gateway parses this query, it first makes a request to devicegqlserver. It adds all the fields that exist on device server, and when it sees a field it doesnt know how to resolve it replalces it with the field func `__federation` and adds theh deviceKeys are fields nested on `__federation`.

Reviewer: ```suggestion
When the gateway parses this query, it first makes a request to device-gqlserver. It adds all the fields that exist on device server. When it sees a field it doesn't know how to resolve, it replaces it with the field func `__federation` and adds the deviceKeys fields nested under `__federation` field.
```

Reviewer: I'm a bit confused about `device server` on this sentence.  Can I get some clarification?

Author: Ah the device server is meant to be `devicegqlserver`, but code-wise it is the gqlserver defined in `federationexample/devicegqlserver/main.go`

```
{
  device {                  #devicegqlserver
    id                      #devicegqlserver
    location {              #devicegqlserver
      latitude              #devicegqlserver
      longitude             #devicegqlserver
      street                #devicegqlserver
    }
    _federation {
        id
        orgId

>     }

Reviewer: This bit is also confusing.  I assume a few things
 - `_federation` -> `__federation` per above description?
 - Which `id` and `orgId` are there referring to from the original query in this example (camera.id and camera.orgId, respectively?)

Author: * We had to change it to one underscore to match the apollo schema validation. The double underscore is reserved for fields used for introspection
* This would be referring to the id and org id on the device. I also added this to the description above

  }
}

```

> Once it gets the repsonse, it makes another request to safetygqlserver to resolve the camera field. The args `deviceKeys` are a list of he `DeviceKey` struct with the id and org Id populated from theh results of the previous query. The field `Device-safetygqserver` runs the function passed into `FetchObjectFromKeys` in the object registration, telling it how to get the full device from the keys. 

Reviewer: ```suggestion
Once the gateway receives the response from devicegqlserver, it forwards another sub-query to safety-gqlserver to resolve the camera fields. The args `deviceKeys` are a list of `DeviceKey` struct with `id` and `orgId` populated from the results of the previous sub-query to devicegqlserver. The field `Device_safetygqserver` runs the function passed into `FetchObjectFromKeys` in the object registration, telling it how to get the full device from the keys. 
```

Reviewer: > The field `Device_safetygqserver` runs the function passed into `FetchObjectFromKeys` in the object registration, telling it how to get the full device from the keys. 

I'm a bit confused about this statement.

Author: The field `Device_safetygqserver` executes the `FetchObjectFromKeys` function defined on the device object. It takes a list of `deviceKeys` and returns a list of  `Device` objects.

> SafetyGqlserver can resolve all the fields but street name, so it once again replaces `street` with the `__federation` field func and adds the keys latitude and longtitudde as fields. 

Reviewer: ```suggestion
Safetygqlserver can resolve all the fields but street name, so it once again replaces `street` with the `__federation` field and adds the keys `latitude` and `longitude` as fields. 
```

```
{

>     _federation {

Reviewer: ```suggestion
    __federation {
```

        Device_safetygqlserver(args: deviceKeys) {
            camera{                    #safetygqlserver
                id                     #safetygqlserver
                isOn                   #safetygqlserver
                isMultiCam             #safetygqlserver
                cameraLocation {       #safetygqlserver
                    latitude           #safetygqlserver
                    longitude          #safetygqlserver

>                     _federation {

Reviewer: ```suggestion
                    __federation {
```

                        latitude
                        longitude
                    }
                }
            }
        }
    }
}
```

> Once the gqlgateway gets the result from the query to safetygqlserver, it makes one more request to deviceservice to resolve street. It sends the locationKeys as arguments and resolve the field street. 

Reviewer: ```suggestion
Once the gateway gets the result from the sub-query to safetygqlserver, it makes one more request to deviceservice to resolve `street` field. It sends the `locationKeys` as arguments.
```

```
{

>     _federation {

Reviewer: ```suggestion
    __federation {
```

Author: We had to change it to one underscore to match the apollo schema validation. The double underscore is reserved for fields used for introspection

        Location_devicegqlserver(args: locationKeys) {
            street
        }
    }
}
```

> One the gqlgateway gets all the results, it stitches it back into a single query

Reviewer: ```suggestion
After all fields from the original query are resolved, the gateway stitches the results into a single response to conform to request query shape.
```

## How to run the example

> Run the bash script `./runfederationexample.sh` and go to the url `http://localhost:3030/federationexample`. You should be able to query all the fields in this example andd try out other things as you modify the example. 

Reviewer: ```suggestion
Run the bash script `./runfederationexample.sh` and go to the url `http://localhost:3030/federationexample`. You should be able to query all the fields in this example and try out other things as you modify the example. 
```

This device gqlserver schema can be found in `federationexample/devicegqlserver/main.go` and the safety gqlserver schema can be found `federationexample/safetyservice/main.go`. 

Let's say we want to run this query. The comment tells us which gqlserver is running the field func. 
```
{
  device {                  #devicegqlserver
    camera{                 #safetygqlserver
      cameraLocation {      #safetygqlserver
        street              #devicegqlserver
   	  }
    }
  }
}
```

When the gateway parses this query, it first makes a request to devicegqlserver. It adds all the fields that exist on devicegqlserver, and when it sees a field it doesn't know how to resolve, it replaces it with the field func `_federation` and adds the deviceKeys (id and orgId) fields nested under `_federation` field.

```
{
  device {                  #devicegqlserver

>     _federation {

Reviewer: ```suggestion
    _federation {.    # this replaces camera field with fields from deviceKeys
```

>         id

Reviewer: where is this id come from, in absolute term.

>         id

Reviewer: ```suggestion
        id.  # refereced to deivceKey.Id
```

# Graphql Federation

In our current world, we have a single graphql schema that contains all the objects and field funcs that can be executed. However, as the number of objects and field funcs scale, we want to allow product teams to each have their own gqlserver that contains the portion of the gqlschema relevant to the product team.

> Federation helps by creating a service called the gqlgateway that is aware of every team gqlserver, which team it belongs to, and the assocaited schema on the gqlserver. Queries are parsed at the main gateway and broken down to sub-queries, which are forwarded to the appropriate gqlservers for execution. The sub-queries only contains field funcs that the team's gqlserver are able to resolve. After receiving all results from the sub-queries, the main gateway stitches all the results into a single result to respond to the original query from the client. This ensures the client can query any field while the schema is federated under the hood. 

Reviewer: ```suggestion
Federation helps by creating a service called the gqlgateway that is aware of every team gqlserver, which team it belongs to, and the assocaited schema on the gqlserver. Queries are parsed at the main gateway and broken down to sub-queries, which are forwarded to the appropriate gqlservers for execution. The sub-queries only contains field funcs that the team's gqlserver are able to resolve. After receiving all results from the sub-queries, the main gateway stitches all the sub-query results into a single result to respond to the original query from the client. This ensures the client can query any field while the schema is federated under the hood. 
```

A non-federated object looks something like this example below 
```
type Device struct {
    Id            int64
    OrgId         int64
    Name          string
}
device := schema.Object("Device", Device{})
```

A federated object has an extra field called FetchObjectFromKeys that takes a function, telling the gqlserver how to fetch the full object using federated key; the minimum number of fields required to fetch the object. A federated object can exist on multiple gqlservers and have different field funcs on each gqlserver.

```
type Device struct {
    Id            int64
    OrgId         int64
    Name          string
}
// Device keys has to be a subset of the fields on Device and is the minimum information needded to fetch a full Device
type DeviceKeys struct {
    Id            int64
    OrgId         int64
}
device := schema.Object("Device", Device{}, schemabuilder.FetchObjectFromKeys(func(args struct{ Keys []*DeviceKeys }) []*Device {
    // Fetch device from database using device keys    
    return fetchedDevices
}))
```

## Example

> At samsara we create iot devices used for vehicle monitoring. One of the devices is a camera owned by our safety team, but the other types of devices don't belong to safety. In this example, we want to have 2 gqlservers, devicegqlserver and safetygqlserver. The safetygqlserver is suppose to have all the field funcs pertaining to the camera. 

Reviewer: ```suggestion
At samsara we create iot devices used for vehicle monitoring. One type of the devices is a camera owned by our safety team, but the other types of devices don't belong to safety team. In this example, we want to have 2 gqlservers, devicegqlserver and safetygqlserver. The safetygqlserver is suppose to have all the field funcs pertaining to the camera. 
```

> At samsara we create iot devices used for vehicle monitoring. One type of the devices is a camera owned by our safety team, but the other types of devices don't belong to the safety team. In this example, we want to have 2 gqlservers, devicegqlserver and safetygqlserver. The safetygqlserver is suppose to have all the field funcs pertaining to the camera. 

Reviewer: ```suggestion
At samsara we create IoT devices used for vehicle monitoring. One type of the devices is a camera owned by our safety team, but the other types of devices don't belong to the safety team. In this example, we want to have 2 gqlservers, devicegqlserver and safetygqlserver. The safetygqlserver is suppose to have all the field funcs pertaining to the camera. 
```

> To do this, on safety serve we add a field called `camera`

Reviewer: ```suggestion
To do this, on safety server we add a field called `camera`
```

```
device.FieldFunc("camera", func(ctx context.Context, device *Device) ([]*Camera , error) {
		return []*Camera{&Camera{Id: 1, IsOn: true}, &Camera{Id: 2, IsOn: true}}, nil
	})
```

> And then register the `camera` object and the field func `cameraLocation`

Reviewer: ```suggestion
We then register the `camera` object and the field func `cameraLocation`
```

``
type Camera struct {
  Id int64
  IsOn bool
}
camera := schema.Object("Camera", Camera{})
camera.FieldFunc("cameraLocation", func(ctx context.Context, camera *Camera) (*Location , error) {
  return &Location{Latitude: 246.245, Longitude: 135.135}, nil
})
```

> This device gqlserver schema can be found in `federationexample/devicegqlserver/main.go` and the safety gqlserver schema can be found `federationexample/safetyservice/main.go`. 

Reviewer: ```suggestion
This devicegqlserver schema can be found in `federationexample/devicegqlserver/main.go` and the safetygqlserver schema can be found `federationexample/safetyservice/main.go`. 
```

> Let's say we want to run this query. The comment tells us which gqlserver is running the field func. 

Reviewer: ```suggestion
Let's run a following query. The comment indicates which gqlserver is resolving each field. 
```

```
{
  device {                  #devicegqlserver
    camera{                 #safetygqlserver
      cameraLocation {      #safetygqlserver
        street              #devicegqlserver

>    	  }

Reviewer: ```suggestion
      }
```

```
{
  device {                  # devicegqlserver
    _federation {           # replaces camera field
        id                  # device key 
        orgId               # device key 
    }
  }
}

```

Once the gateway receives the response from devicegqlserver, it forwards another sub-query to safetygqlserver to resolve the `camera` and `cameraLocation` field. 

The args `deviceKeys` are a list of `DeviceKey` structs with `id` and `orgId` populated from the results of the previous sub-query to devicegqlserver. The field `Device_safetygqserver` executes the `FetchObjectFromKeys` function defined on the device object. It takes a list of `deviceKeys` and returns a list of  `Device` objects.

Safetygqlserver can resolve all the fields but street name, so it once again replaces `street` with the `_federation` field and adds the keys `latitude` and `longitude` as fields. 

```
{
    _federation {
        Device_safetygqlserver(args: deviceKeys) {
            camera{                    # safetygqlserver
                cameraLocation {       # safetygqlserver

>                     _federation {      # eplaces 

Reviewer: ```suggestion
                    _federation {      # replaced
```
